<a href="https://colab.research.google.com/github/m-dergacheva/Yandex-Practicum_DA/blob/main/SQL_%D0%9A%D0%BD%D0%B8%D0%B6%D0%BD%D1%8B%D0%B9_%D0%BC%D0%B0%D0%B3%D0%B0%D0%B7%D0%B8%D0%BD.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Мини-кейс SQL:Книжный магазин

## Описание:
Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг.

Цель: обозначить паттерны, чтобы сформулировать ценностное предложение для нового продукта.

In [ ]:
# импортируем библиотеки
import pandas as pd
import sqlalchemy as sa

In [ ]:
# устанавливаем параметры
db_config = {
'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db' # название базы данных
}
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)
# сохраняем коннектор
engine = sa.create_engine(connection_string, connect_args={'sslmode':'require'})
# чтобы выполнить SQL-запрос, пишем функцию с использованием Pandas
def get_sql_data(query:str, engine:sa.engine.base.Engine=engine) -> pd.DataFrame:
  with engine.connect() as con:
    return pd.read_sql(sql=sa.text(query), con = con)

### Исследуйте таблицы — выведите первые строки

In [ ]:
# формируем запрос и выводим данные таблицы books
query_books = '''SELECT * FROM books LIMIT 5'''
get_sql_data(query_books)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

В таблице books содержатся данные о id книг, id авторов, наименование книг, кол-во страниц, даты публикации и id издательства, выпустившее книгу.

In [ ]:
# формируем запрос и выводим данные таблицы authors
query_authors = '''SELECT * FROM authors LIMIT 5'''
get_sql_data(query_authors)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

В таблице authors содержатся данные о id авторов и ФИО этих авторов

In [ ]:
# формируем запрос и выводим данные таблицы ratings
query_ratings = '''SELECT * FROM ratings LIMIT 5'''
get_sql_data(query_ratings)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

В таблице ratings содержатся данные о id оценки, id книг, ник-нейм пользователя и оценка книги, которую присвоил пользователь

In [ ]:
# формируем запрос и выводим данные таблицы reviews
query_reviews = '''SELECT * FROM reviews LIMIT 5'''
get_sql_data(query_reviews)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

В таблице reviews содержатся данные о id обзора, id книги, к которой написан обзор, ник-нейм пользователя, оставивший этот обзор, и текст обзора

In [ ]:
# формируем запрос и выводим данные таблицы publishers
query_publishers = '''SELECT * FROM publishers LIMIT 5'''
get_sql_data(query_publishers)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [ ]:
query_publishers1 = '''SELECT *, split_part(publisher, ' ', 2) FROM publishers LIMIT 5'''
get_sql_data(query_publishers1)

publisher_id                          publisher split_part
0             1                                Ace           
1             2                           Ace Book       Book
2             3                          Ace Books      Books
3             4                      Ace Hardcover  Hardcover
4             5  Addison Wesley Publishing Company     Wesley

В таблице publishers содержатся данные о id издательств и наименование издательств

### Исследуйте таблицы — посчитайте количество строк в каждой таблице

In [ ]:
# формируем запрос и выводим данные таблицы books
count_query_books = '''SELECT count(*) FROM books'''
get_sql_data(count_query_books)

count
0   1000

Количество строк в таблице books 1000 строк

In [ ]:
# формируем запрос и выводим данные таблицы authors
count_query_authors = '''SELECT count(*) FROM authors'''
get_sql_data(count_query_authors)

count
0    636

Количество строк в таблице authors 636 строк

In [ ]:
# формируем запрос и выводим данные таблицы ratings
count_query_ratings = '''SELECT count(*) FROM ratings'''
get_sql_data(count_query_ratings)

count
0   6456

Количество строк в таблице ratings 6456 строк

In [ ]:
# формируем запрос и выводим данные таблицы reviews
count_query_reviews = '''SELECT count(*) FROM reviews'''
get_sql_data(count_query_reviews)

count
0   2793

Количество строк в таблице reviews 2793 строк

In [ ]:
# формируем запрос и выводим данные таблицы publishers
count_query_publishers = '''SELECT count(*) FROM publishers'''
get_sql_data(count_query_publishers)

count
0    340

Количество строк в таблице publishers 340 строк

### Посчитайте, сколько книг вышло после 1 января 2000 года

In [ ]:
#Посчитайте, сколько книг вышло после 1 января 2000 года
count_book='''
SELECT COUNT(book_id) AS count_book
FROM books
WHERE publication_date> '2000-01-01'
'''
get_sql_data(count_book)

count_book
0         819

После 1 января 2000 года было выпущено 819 книг

### Для каждой книги посчитайте количество обзоров и среднюю оценку

In [ ]:
#Для каждой книги посчитайте количество обзоров и среднюю оценку
avg_rating='''
SELECT
  books.title AS book_title,
  books.book_id AS book_id,
  AVG(ratings.rating) AS rating_avg,
  COUNT(distinct reviews.review_id) AS count_rewiews
FROM books AS books
INNER JOIN ratings AS ratings ON books.book_id=ratings.book_id
Left JOIN  reviews AS reviews ON books.book_id=reviews.book_id
GROUP BY books.title,
  books.book_id
having COUNT(distinct reviews.review_id) = 6 or COUNT(distinct reviews.review_id) = 7
ORDER BY count_rewiews DESC, rating_avg DESC
'''
get_sql_data(avg_rating)

book_title  book_id  rating_avg  \
0                             Twilight (Twilight  #1)      948    3.662500   
1   Harry Potter and the Prisoner of Azkaban (Harr...      302    4.414634   
2   Harry Potter and the Chamber of Secrets (Harry...      299    4.287500   
3                                      The Book Thief      656    4.264151   
4                                    The Glass Castle      734    4.206897   
5                           Outlander (Outlander  #1)      497    4.125000   
6                 The Hobbit  or There and Back Again      750    4.125000   
7   The Curious Incident of the Dog in the Night-Time      695    4.081081   
8   The Lightning Thief (Percy Jackson and the Oly...      779    4.080645   
9                                 Water for Elephants      963    3.977273   
10             The Da Vinci Code (Robert Langdon  #2)      696    3.830508   
11                             The Catcher in the Rye      673    3.825581   
12                                      The Alchemist      627    3.789474   
13                                           The Road      854    3.772727   
14                          The Giver (The Giver  #1)      733    3.750000   
15                                    Eat  Pray  Love      207    3.395833   

    count_rewiews  
0               7  
1               6  
2               6  
3               6  
4               6  
5               6  
6               6  
7               6  
8               6  
9               6  
10              6  
11              6  
12              6  
13              6  
14              6  
15              6

Только у одной книги Twilight (Twilight 1) максимальное кол-во обзоров - 7. Самый высокий рейтинг соответствует в основном книгам, кол-во обзоров на которых составляет 1-2 (за исключением, например, A Dirty Job (Grim Reaper #1), у которой кол-во обзоров 4). Предполагается, что высокий рейтинг обоснован малым кол-вом обзоров но с высокими рейтингами.

### Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры

In [ ]:
#Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры
publisher_50='''
SELECT publisher.publisher as name_publisher,
  count(book.title) as books_name_count
FROM books as book
INNER JOIN publishers as publisher on book.publisher_id=publisher.publisher_id
WHERE book.num_pages>50
GROUP BY publisher.publisher_id
ORDER BY books_name_count desc
LIMIT 15
'''
get_sql_data(publisher_50)

name_publisher  books_name_count
0               Penguin Books                42
1                     Vintage                31
2    Grand Central Publishing                25
3            Penguin Classics                24
4            Ballantine Books                19
5                      Bantam                19
6                     Berkley                17
7          St. Martin's Press                14
8               Berkley Books                14
9   William Morrow Paperbacks                13
10                      Delta                13
11              HarperCollins                12
12  Little  Brown and Company                12
13           Harper Perennial                12
14                  Tor Books                11

Издательство, выпустившее максимальное кол-во книг, толще 50 страниц, - Penguin Books

### Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

In [ ]:
#Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

author_max_rat_50='''
select temp.name_author as author_name,
  avg(temp.rating) as rating_avg
from (select authors.author as name_author,
  avg(ratings.rating) as rating,
  books.book_id as book,
  count(ratings.rating) as book_rating
from books as books
inner join authors as authors on books.author_id=authors.author_id
inner join ratings as ratings on books.book_id=ratings.book_id
group by name_author,
  book
having count(ratings.rating)>=50) as temp
group by temp.name_author
order by rating_avg desc
limit 15
'''
get_sql_data(author_max_rat_50)

author_name  rating_avg
0                          J.K. Rowling/Mary GrandPré    4.283844
1                   Markus Zusak/Cao Xuân Việt Khương    4.264151
2                                      J.R.R. Tolkien    4.258446
3                                   Louisa May Alcott    4.192308
4                                        Rick Riordan    4.080645
5                                     William Golding    3.901408
6                                       J.D. Salinger    3.825581
7            Paulo Coelho/Alan R. Clarke/Özdemir İnce    3.789474
8   William Shakespeare/Paul Werstine/Barbara A. M...    3.787879
9                                           Dan Brown    3.754540
10                                         Lois Lowry    3.750000
11            George Orwell/Boris Grabnar/Peter Škerl    3.729730
12                                    Stephenie Meyer    3.662500
13                                     John Steinbeck    3.622951

Самая высокая средняя оценка соответствует автору J.K. Rowling/Mary GrandPré, чья средняя оценка книг - 4.283844

### Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок

In [ ]:
#Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок
count_rewiews_users ='''
select avg(temp2.text) as avg_count_text
from (select count(reviews.text) as text,
  temp.username as username
from reviews as reviews
join (select count(ratings.rating_id) as count_rating,
  ratings.username as username
from ratings
group by ratings.username
having count(ratings.rating_id)>48) as temp on reviews.username=temp.username
group by temp.username) as temp2
'''
get_sql_data(count_rewiews_users)

avg_count_text
0            24.0

Среднее количество обзоров от пользователей, которые поставили больше 48 оценок составляет 24.

## Выводы

1. На среднюю оценку пользователей сильно влияет непосредственно количество таких оценок: чем меньше на книгу выпущено обзор, тем менее "реальной" будет оценка, так, например, в целом на книги с максимальным рейтингом выпущено по 1-2 обзора.
2. Наиболее популярное издательство, выпускающее книги толще 50 стр. (не брошюры) - Penguin Books. На 2-4 местах в ТОП 5 находятся издания: Vintage, Grand Central Publishing, Penguin Classics и Bantam.  
3. В ТОП-5 авторов, чей рейтинг находится в районе 4-5 баллов: J.K. Rowling/Mary GrandPré,Markus Zusak/Cao Xuân Việt Khương, J.R.R. Tolkien, Louisa May Alcott и Rick Riordan. Предполагается, что такие авторы пишут наиболее понравившиеся аудитории книги.  
4. Среднее количество обзоров пользователей, которые активно голосуют, составляет 24 отзыва. Предполагается, что отзывам таких читателей стоит "прислушиваться", т.к. они лучше разбираются в литературе (им есть с чем сравнить).  

### Рекомендации

1. Для потенциальной продажи книг, предполагается, что наилучшим вариантом будет предлагать книги с большим количеством оценок, т.к. средняя оценка пользователей на такие книги будет более "приближена к реальности".  
2. Предполагается, что отзывам читателей, которые написали среднее и выше кол-во обзоров (более или равно 24 кол-во) стоит "прислушиваться", т.к. они лучше разбираются в литературе (им есть с чем сравнить)
3. Для продажи книг, вероятно, следует допускать/выгружать в первую очередь книги, которые пишут авторы с высокими рейтингами (4-5 баллов)